Replace bootstrap-sass 3 to bootstrap 4.6.1

tnir opened the pull request:
| before (1024-width on iPad Pro) | after (1024-width on iPad Pro) |
|-|-|
| ![bundler io_(iPad Pro)](https://user-images.githubusercontent.com/10229505/177666238-49c2f0b2-54b1-4416-9f87-3a7cecedd658.png) | ![bundler-site-tnir-boots-dje5be herokuapp com_(iPad Pro)](https://user-images.githubusercontent.com/10229505/178105680-35b23f69-7671-439f-8467-05da1b1fb9eb.png) |

Closes #643

### What was the end-user problem that led to this PR?

An end-user needs to load a legacy Bootstrap 3 just to browse the bundler docs.

### What was your diagnosis of the problem?

No maintainers/contributors maintained the frontend of the site for 6 years.

### What is your fix for the problem, implemented in this PR?

HAML and SCSS would be easier for everyone to maintain as Bootstrap 4 is more recent version than 3.

#### Screenshots for navbar/searchbox
![comment-1177209898](https://user-images.githubusercontent.com/10229505/177733076-694273d0-44d0-4dde-bf2f-2d0ee18b5e1c.png)

#### New introduction
- https://fontawesome.com/v5/icons/search?s=solid (from Font Awesome 5 Free) (dropped Glyphicons icon font in Bootstrap 4)
- Opacity utilities from Bootstrap 5 https://getbootstrap.com/docs/5.1/utilities/colors/#opacity
  - which will be removed when Bootstrap upgraded from 4.6 to 5.1 (or upcoming 5.2)

#### Small improvement in searchbox spacing

##### Before this PR

https://user-images.githubusercontent.com/10229505/177769132-cdef09f2-f5c7-4a25-8cd9-67a4226dcfc8.mp4

##### After this PR
https://user-images.githubusercontent.com/10229505/177769152-e03cc14d-5681-40d6-b63b-ce1596ce312c.mp4

### Why did you choose this fix out of the possible options?

I cannot upgrade it from 3 to 5 (the latest major version as of creation of the PR) directly.

- Updates lots from #218

Signed-off-by: Takuya Noguchi [takninnovationresearch@gmail.com](https://github.com/sponsors/tnir)

tnir commented on source/v2.3/docs.html.haml:
```diff
@@ -6,8 +6,8 @@
 
-.row.docs
-  .container
+.container
+  .row.docs
```
Generally speaking, the original code is not standard, but the fix is Boostrap-standard. But we need to change all occurrences in all `source/v*/docs.html.haml` files.

With some changes done in `source/layouts/commands_layout.haml`, this change is not mandatory as of this PR:
```suggestion
.row.docs
  .container
```

tnir replied:
Fixed in db54cc7.

tnir replied:
db54cc7 was wrong...

tnir commented on source/shared/_whats_new.haml:
```diff
@@ -2,4 +2,4 @@
   .col-sm-12
     - versions.reverse.select{ |version| path_exist?('whats_new', version) }.each do |v|
-      = link_to v, "/#{v}/whats_new.html", class: "btn btn-sm #{current_visible_version == v ? 'btn-default' : 'btn-primary'}"
+      = link_to v, "/#{v}/whats_new.html", class: "btn btn-sm #{current_visible_version == v ? 'btn-secondary' : 'btn-primary'}"
```
`btn-secondary` (Bootstrap 4; BS4) is certainly the replacement of `btn-default` (BS3), but the appearance changes. 🤔 

cf. https://getbootstrap.com/docs/4.0/migration/#buttons

So, the color can be swapped at this time like:
```suggestion
      = link_to v, "/#{v}/whats_new.html", class: "btn btn-sm #{current_visible_version == v ? 'btn-primary' : 'btn-secondary'}"
```

tnir replied:
Filed as #645

tnir replied:
This was dissolved in #873.

tnir commented on assets/stylesheets/_footer.scss:
```diff
@@ -12,15 +12,12 @@ body {
 .footer {
-  position: absolute;
-  bottom: 0;
+  // position: absolute;
+  // bottom: 0;
```
```suggestion
```
Forgot to remove these lines commented out.

tnir replied:
Fixed in 487b0b8.

tnir commented on assets/stylesheets/_docs.scss:
```diff
@@ -32,3 +32,8 @@
   }
 }
+
+pre {
+  padding: 0.625rem;
+  border-radius: 4px;
+}
```
Transitional code from BS3 to BS4.

tnir commented on assets/javascripts/application.js:
```diff
@@ -1,9 +1,8 @@
-import 'bootstrap-sass/assets/javascripts/bootstrap/transition'
```
transitons were written in SCSS in BS4.

tnir commented on source/contributors.html.haml:
```diff
@@ -5,7 +5,7 @@ title: Contributors
   = image_tag '/images/team_header_transparent_bg.png',
     srcset: '/images/team_header_transparent_bg.png 1x, /images/team_header_transparent_bg@2x.png 2x, /images/team_header_transparent_bg@3x.png 3x',
-    class: 'img-responsive header-padding',
+    class: 'img-fluid header-padding',
```
We needed to include `image` SCSS from `application.css.scss`.

```
@import "~bootstrap/scss/image";
```

tnir replied:
```scss
@import "~bootstrap/scss/images";
```

tnir replied:
Done in d68421a

duckinator commented:
I can't give this a proper review right now, but I just wanted to say thank you _so much_ for taking this on. It is greatly appreciated. :heart: 

If nobody else gives this PR a proper review by next Monday (the 11th), feel free to @-mention me and I'll take a closer look at it.

tnir commented:
@olleolleolle Welcome and I just miss Turbinkanalen.

@duckinator Thanks a lot in advance for reviewing my upcoming PR for updating Bootstrap 4 to 5 or removal of it (not decided yet) hopefully I would create around July 11.

tnir commented on source/layouts/_navbar.haml:
```diff
@@ -1,21 +1,15 @@
-        %li= link_to t('navbar.blog'), '/blog'
-        %li= link_to t('navbar.repository'), 'https://github.com/rubygems/rubygems/tree/master/bundler', target: '_blank', rel: 'noopener noreferrer'
+%link{rel: "stylesheet", href: "https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/all.min.css", integrity: "sha512-1ycn6IcaQQ40/MKBW2W4Rhis/DbILU74C1vSrLJxCq57o941Ym01SwNsOMqvEBFlcgUa6xLiPY/NS5R+E6ztJQ==", crossorigin: "anonymous", referrerpolicy: "no-referrer"}
+.container
+  %nav.navbar.navbar-expand-lg.navbar-light.justify-content-between
```
Looks like this can be acceptable:
```suggestion
%nav.navbar.navbar-expand-md.navbar-light.container.justify-content-between
```

tnir replied:
Fixed it in 058b6c9.

tnir commented on source/layouts/_footer.haml:
```diff
@@ -1,7 +1,7 @@
-      %li= link_to t('navbar.team'), "/contributors.html"
-      %li= link_to t('navbar.blog'), "/blog"
-      %li= link_to t('navbar.repository'), "https://github.com/rubygems/rubygems/tree/master/bundler"
+.container.text-center
+  %nav.navbar.navbar-expand-lg.justify-content-center
```
```suggestion
%nav.navbar.navbar-expand-sm.justify-content-center
```

tnir replied:
Fixed it in 058b6c9.

tnir reviewed (commented):
You can improve two more structures around [navbar](https://getbootstrap.com/docs/4.6/components/navbar/).

tnir commented:
| before (1024-width on iPad Pro) | after (1024-width on iPad Pro) |
|-|-|
| ![bundler io_(iPad Pro)](https://user-images.githubusercontent.com/10229505/177666238-49c2f0b2-54b1-4416-9f87-3a7cecedd658.png) | ![bundler-site-tnir-boots-dje5be herokuapp com_(iPad Pro)](https://user-images.githubusercontent.com/10229505/177666227-8058b7d2-ff5f-492e-9fd6-9ce654550710.png) |

As of 058b6c9, `.pull-right`s were not replaced with `float-right`.

deivid-rodriguez commented:
Loving this! I looked at homepages side by side and the differences look acceptable, but I do like the current nabber better with a more spacious search bar and the gray search icon.

tnir commented:
Thanks @deivid-rodriguez for the feedback.

> [...] I do like the current navbar better with a more spacious search bar and the gray search icon.

In my previous changeset, while I felt cramped against the width, I (somehow intentionally) dropped the magic number width `300px` just by leaving ourselves to a browser they is using (the default width looks like 20 characters).

But as per your comment, I just updated the change in the PR. Also the opacity for the icon was reverted in the latest commit:

![comment-1177209898](https://user-images.githubusercontent.com/10229505/177733076-694273d0-44d0-4dde-bf2f-2d0ee18b5e1c.png)

deivid-rodriguez commented:
Thanks, but now the icon seems a bit off and falls outside of the search box? Feels nicer as it is now. I understand it's tricky to replicate the current appearance exactly.

#### Now

<img width="412" alt="Captura de Pantalla 2022-07-07 a las 11 47 55" src="https://user-images.githubusercontent.com/2887858/177745198-4487fb5f-715e-4f78-bf5b-3c36272115a3.png">

#### With this PR

<img width="393" alt="Captura de Pantalla 2022-07-07 a las 11 48 19" src="https://user-images.githubusercontent.com/2887858/177745226-a4379637-068d-4c45-b09a-c8d231386fc1.png">

tnir commented:
Thanks. Added screen videos after your suggestions applied in the `Small improvement in searchbox spacing` section of the PR description above.

deivid-rodriguez commented:
Nice!

I think the final feedback from my side (appearance wise) is that the spacing in the current site feels more relaxed and visually pleasant that with this PR. For example,

#### In the expanded "hamburger menu", items are too closed together and search is not properly aligned

| Before | After |
|--------|------|
|<img width="400" alt="Captura de Pantalla 2022-07-07 a las 14 38 32" src="https://user-images.githubusercontent.com/2887858/177775906-a5df8215-a81b-43f9-9551-b5af5ccd4f28.png"> | <img width="400" alt="Captura de Pantalla 2022-07-07 a las 14 38 12" src="https://user-images.githubusercontent.com/2887858/177776072-a6a856ac-0480-4605-9402-35795cded252.png"> |

#### In the footer, items are too close together and content feels to close to footer in the new site.

| Before | After |
|--------|------|
|<img width="560" alt="Captura de Pantalla 2022-07-07 a las 14 27 46" src="https://user-images.githubusercontent.com/2887858/177773459-f4ffbade-5dd6-461d-b2f8-a7a2068bd3c7.png">|<img width="569" alt="Captura de Pantalla 2022-07-07 a las 14 28 00" src="https://user-images.githubusercontent.com/2887858/177773451-1f5df284-7c1c-4a9c-a386-bb6af3790c65.png">|

tnir commented:
> In the expanded "hamburger menu", items are too closed together [...]

To keep the height of `nav.nav` as original (`52px` (in font 100% env)), I used to make a tweak in global in my previous commit. This was a side effect from that, so now that tweak in global was reverted by the latest commit (`ac10d83`). Instead of the tweak in global, another tweak is added in `assets/stylesheets/_navbar.scss`.

> In the expanded "hamburger menu", [...] search is not properly aligned

Fixed by the latest commit (`ac10d83`)

> In the footer, items are too close together and content feels to close to footer in the new site.

Fixed by the latest commit (`ac10d83`) (e62988ab1af0628cbd8328fe3b90d544080520ea..fb86b74612c4db4ab80751424d052feb9ff80c89).

----

Also to address a broken UI in a limited case, another fix was applied: fb86b74612c4db4ab80751424d052feb9ff80c89..ce00031ca3f561e37927f001b8b3fb9158611835

deivid-rodriguez commented:
Almost perfect, spacing looks perfect all around the site now! I actually recall considering taking this task, and quickly gave up because it felt too much work, now it's almost done! 😍 

Anyways, one last round of details. Every other little visual discrepancy makes no difference in my eyes.

### Still on the search box, the base line seems too far from the icon and the placeholder text. Also, the icon is not properly aligned right and the hamburger icon is maybe a bit too big

|Before|After|
|-|-|
|<img width="560" alt="Captura de Pantalla 2022-07-07 a las 19 07 17" src="https://user-images.githubusercontent.com/2887858/177831461-79043951-a62c-44c0-b10a-5bd890a65e41.png">|<img width="560" alt="Captura de Pantalla 2022-07-07 a las 19 07 05" src="https://user-images.githubusercontent.com/2887858/177831487-ad54a5d4-cfeb-4596-bd1e-1404990944d4.png">|

### Color of links is brighter than the previous color. The previous color looks better to me. Also inline blocks of code no longer get the proper text color and background color.

|Before|After|
|-|-|
|<img width="551" alt="Captura de Pantalla 2022-07-07 a las 19 16 11" src="https://user-images.githubusercontent.com/2887858/177832253-7e3758c1-4d8e-4270-ae4e-95cc1c5c9a42.png">|<img width="559" alt="Captura de Pantalla 2022-07-07 a las 19 16 24" src="https://user-images.githubusercontent.com/2887858/177832237-197c0e69-358d-4294-8f3a-0850c83c6b4f.png">|

tnir commented:
> Still on the search box, the base line seems too far from the icon and the placeholder text. Also, the icon is not properly aligned right and the hamburger icon is maybe a bit too big

While you might refer older change (since the brand text color is different from one in my recent change), all fixed.
| Before | After |
|-|-|
| ![bundler io_](https://user-images.githubusercontent.com/10229505/177898939-8c07ead5-5476-4a95-b479-ac05af2c6ebb.png) | ![bundler-site-tnir-boots-dje5be herokuapp com_](https://user-images.githubusercontent.com/10229505/177898669-b630acff-ef33-4ed2-80cb-19c2d66810ff.png) |
| `42px` pitch in height of `li.nav-item` | `40px` pitch |
| 44x34 (hamburger icon) | (56x40) => 50x34 |

While the height of the hamburger icon was reverted in the latest commit, it looks too small on my mobile device. I guess that is why Bootstrap 4 team enlarged this `navbar-toggler` to 125% between Bootstrap 3 and 4. But less changes are nice **in this PR**.

> Your app's UI is easier to use if it contains controls that are easier to see and tap. We recommend that each interactive UI element have a focusable area, or touch target size, of at least 48dp ✕ 48dp. Larger is even better.
> https://developer.android.com/guide/topics/ui/accessibility/apps#large-controls

---

> Color of links is brighter than the previous color. The previous color looks better to me. Also inline blocks of code no longer get the proper text color and background color.

Code was manually backported from Bootstrap 3.4.

deivid-rodriguez commented:
I'm fine keeping the new hamburger icon size if it renders better for you, we can do it on a separate PR I guess.

I still see some of the little differences I reported.

* The color of links is a bright blue while before it was a darker blue.
* The search icon does not properly float to the right in narrow screens.

* Also noticed that the gray tonality of the search icon and the hamburger icon is lighter (and nicer, IMO) in the current site.

tnir commented:
> I'm fine keeping the new hamburger icon size if it renders better for you, we can do it on a separate PR I guess.

Reverted from 1.25rem (BS4 default and my previous change request) to 1rem (BS3 default and the current implementation).

> The color of links is a bright blue while before it was a darker blue.

Regression happened 🙏 . Fixed it in the latest commit. `a` has now `color: #337ab7`.

> The search icon does not properly float to the right in narrow screens.

I could not reproduce this in my environment (Google Chrome 103, and Firefox 102.0.1 on macOS 12.4). Could you show your environment?

> Also noticed that the gray tonality of the search icon and the hamburger icon is lighter (and nicer, IMO) in the current site.

Fixed it in the latest commit with tweaks. `opacity: 0.4` was created and applied to search icon and `navbar-light-toggler-icon-bg` variable was backported from Bootstrap 3.4 and applied to `.navbar-toggler-icon` element.

deivid-rodriguez commented:
Thanks for all the changes 😍.

Regarding the search icon not being properly floated, I don't see it anymore, seems mostly fine now (I'm testing this in Safari).

Some things I did notice while verifying this though:

* The breakpoint to start collapsing the menu and showing the hamburger icon seem to  have changed from 767px to 991px. Previous breakpoint seems better because the full menu and content are perfectly readable at 767-991 widths, I believe.
* Once the menu collapses, site title and hamburger icon are not properly floated to left and right respectively, but instead leave whitespace margins on each size. But while writing this and observing the behavior, I realized I might like the new behavior better? Not sure, up to you :)
* Hamburger icon doesn't turn gray when hovering.

tnir commented:
> The breakpoint to start collapsing the menu and showing the hamburger icon seem to have changed from 767px to 991px. Previous breakpoint seems better because the full menu and content are perfectly readable at 767-991 widths, I believe.

In the changeset previously suggested, [the search box was requested wider (300px)](https://github.com/rubygems/bundler-site/pull/644#issuecomment-1177209898) so needed to chang to ~ 991px (= `lg`) for a failsafe, but just now I found `md` breakpoint looks no problem at least in my environment.

So fixed it in the latest commit.

> Once the menu collapses, site title and hamburger icon are not properly floated to left and right respectively, but instead leave whitespace margins on each size. But while writing this and observing the behavior, I realized I might like the new behavior better? Not sure, up to you :)

I guess technical debts on this root cause are widely spreaded but I do not wan to spend more time on this. As per requested, I can do it.

So fixed it in the latest commit.

deivid-rodriguez commented:
> I guess technical debts on this root cause are widely spreaded but I do not wan to spend more time on this. As per requested, I can do it only for this point at this moment. I think global changes should be done once https://github.com/rubygems/bundler-site/issues/628 is addressed.

> So fixed it in the latest commit.

I'm not sure if you mean you don't want to spend more time on this PR, or you don't want to spend time on addressing other design improvements (other than keeping the appearance as similar as possible to the current one in this PR).

In any case, I think the current appearance is ok now, but I will leave this PR opened a few more days so that other people with better eye for design can look at it (I think @duckinator wanted to have a look).

This is really great work, thanks!!

tnir commented:
> Hamburger icon doesn't turn gray when hovering.

Totally not sure why this is needed. Again Bootstrap 3 was born in 2013. I guess the current behavior just resulted from the legacy code included in Bootstrap 3. Setting `.navbar-toggle:hover` is a nonsense at least in 2018.

cf. https://github.com/twbs/bootstrap/commit/3e896193abb3e529abd9260ebfc60d284071c4f3

tnir commented:
If you are okay, let us ping @sailorhg @kruczjak for their comments as per https://bundler.io/contributors.html

deivid-rodriguez commented:
> Totally not sure why this is needed. Again Bootstrap 3 was born in 2013. I guess the current behavior just resulted from the legacy code included in Bootstrap 3. Setting .navbar-toggle:hover is a nonsense at least in 2018.

It's not really needed, but if I had to choose I like it better with the hover effect. I think including gentle hover effects on things that are clickable is a nice thing. But it's fine as is too, no worries.

> If you (@deivid-rodrigue) are okay, let us ping @sailorhg @kruczjak for their comments as per https://bundler.io/contributors.html

Sure, but they have not been active for a long time, and I'm not sure they wish to be active again or leave any feedback here 🤷.

tnir commented:
All make sense to me as well.

Comments from anyone else than @deivid-rodriguez are now welcome again.

tnir commented:
/ping @duckinator

duckinator commented:
The only problem I see is that I get an unnecessary horizontal scrollbar, but I don't think that needs to block merging this. Aside from that this PR is good to go!

The scrollbar appears to be caused by the `row` class, and everything with that class is wider than my screen. Removing `margin-right: -15px` on `.row` fixes it, but I don't know if it's the actual cause of the problem:

![image](https://user-images.githubusercontent.com/39698/178346937-8163e3a1-65b5-4d74-aac6-b8f5c9571f79.png)

![image](https://user-images.githubusercontent.com/39698/178347004-a74f5dbb-6fc5-479b-a1c1-2edee0cde8e8.png)


@tnir would you rather I merge this as-is and open an issue for the scrollbar thing, or do you want to try to include the fix for that here? imo it's minor enough that I don't mind merging it and fixing that later.

tnir commented:
@duckinator Good catch! While the original code looks non-standard, I fixed it in the latest commit so I think this would be ready to merge.

duckinator commented:
This looks good to me, so I'm merging it.

Thank you so much for all of your work on the site, @tnir!

tnir commented:
Thanks @duckinator for a quick merging. This has blocked many tasks for me.

deivid-rodriguez commented:
I'm getting bad styles in what's new page, I guess it was after this PR.

<img width="1438" alt="Captura de Pantalla 2022-07-12 a las 14 01 08" src="https://user-images.githubusercontent.com/2887858/178485288-d4cc494a-e7c1-46bf-809b-9d43c1bf08b6.png">

tnir commented:
@deivid-rodriguez Yeah, I found this regression in last minute changes at https://github.com/rubygems/bundler-site/pull/644#event-6973239125 soon after this PR got merged.

deivid-rodriguez commented:
Thanks for fixing it!
